src/cogent3/core/new_alignment.py

```diff
@@ -4320,6 +4320,38 @@ def filtered(
 
         return self.gapped_by_map(keep, info=self.info)
 
+    def no_degenerates(self, motif_length: int = 1, allow_gap: bool = False):
+        """returns new alignment without degenerate characters
+
+        Parameters
+        ----------
+        motif_length
+            sequences are segmented into units of this size and the segments are
+            excluded if they contain degenerate characters.
+        allow_gap
+            whether gaps are allowed or whether they are treated as a degenerate
+            character (latter is default, as most evolutionary modelling treats
+            gaps as N).
+        """
+        # refactor: array
```

Comment:
note, the alphabets are designed such that the canonical characters occur before the first gap, then the first ambiguity etc.. So, an array approach is

```python
# make sure valid moltype, then
cutoff = len(self.moltype)
indices = (self.array_pos < cutoff).all(axis=1)
# for when motif_length != 1, but already handled non-modulo sequence
num_motif = len(self) // motif_length
motif_valid = indices.reshape(num_motif, motif_length).any(axis=1).flatten()
selected = self.array_pos[selected]
```
or something like that

Reply:
@GavinHuttley this makes me think that we should have a from_array method on `AlignedSeqsData`, where we provide the array of sequences and a list of names defining the order of the sequences in the array. What do you think

```py

@classmethod
def from_names_and_array(names: list, data: numpy.ndarray) -> AlignedSeqsData:
    ...

```

Reply:
agreed